docs/proposals/cri.md

```diff
@@ -0,0 +1,313 @@
+
+### Goals
+CRI support in edged must:
+* support multiple runtimes like docker, contianerd, crio etc.
+* Support corresponding CNI with pause container and IP
+
+### Non-goals
+
+* Automatic detection of container runtimes and its selection.
+* 
+## Proposal
+
+Currently Kubernetes kubelet CRI supports container runtimes like containerd, crio etc but support for docker runtime is
+provided using dockershim as well. However going forward even docker runtime will be supported through only CRI. Also 
+currently kubeedge edged supports only docker runtime using the legacy dockertools. Hence we propose to support multiple 
+container runtime in kubeedge edged as follows
+1. Include CRI support as in kubernetes kubelet to support contianerd, crio etc
+2. Continue with docker runtime support using legacy dockertools until CRI support for the same is available i.e. support
+for docker runtime using dockershim is not considered in edged
+
+
+### Use Cases
+
+* Customer can choose the required runtime on his edge platform
+
+
+## High Level Design
+
+### Edged with CRI support
+<img src="../docs/images/edged/docker-cri.png">
+
+## Low Level Design
+
+### Packages
+
+Additional CRI runtime packages need to be included.
+
+```go
+"k8s.io/kubernetes/pkg/kubelet/kuberuntime"
+"k8s.io/kubernetes/pkg/kubelet/remote"
+
+```
+
+### Configuration parameters
+
+The options to configure runtime type i.e. docker or other remote runtimes(crio, containerd etc) and the CRI remote endpoints
+for runtime and image services should be provided.
+
+```go
+type Config struct {
+       ....
+       runtimeType           string
+       remoteRuntimeEndpoint string
+       remoteImageEndpoint   string
+       RuntimeRequestTimeout metav1.Duration
+       ....
+ }
+ ```
+### Data structure modifications
+
+The edged data strcuture needs to include the generic runtime which currently has docker runtime including the runtime state and
+runtime name.
+
+```go
+//Define edged
+type edged struct {
+  ....
+  // Last timestamp when runtime responded on ping
+  // Mutex is used to protect this value
+  runtimeState *runtimeState
+  containerRuntimeName string
+  // Container runtime
+  containerRuntime kubecontainer.Runtime
+  // updateRuntimeMux is a lock on updating runtime, because this path is not thread-safe
+  // This lock is used by Kubelet.updateRuntimeUp function and shouldn't be used anywhere else
+  updateRuntimeMux sync.Mutex
+  ....
+}
+
+```
+
+###  Edged object creation modifications
+
+The existing newEdged() function needs to modified include creating CRI runtime object based on the runtime type including
+creations of objects for runtime and image services. However the existing edged does not provide the support for all the 
+parameters required to create the CRI runtime object and default parameters need to be considered for the same.
+
+```go
+
+//newEdged creates new edged object and initialises it
+func newEdged() (*edged, error) {
```

Comment:
I feel it's improper to give some many codes here, pseudocode instead please.

Reply:
Understand but easier to specify code than pseudo code :-)

Reply:
Used pseudocode